### PR TITLE
feat(formula): add logical functions

### PR DESCRIPTION
**Pull request by Hossein-Khan**
### Summary

This PR adds implementations for the following logical functions:
- IFNA
- IFS
- NOT
- XOR
- FALSE
- TRUE
- SWITCH

### Changes
- Added new classes for each logical function in the appropriate directory.
- Ensured each function follows the existing structure and conventions.
- Added unit tests for each new function.

### CODE REVIEW

File: packages/engine-formula/src/functions/logical/ifna/__tests__/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        });
+
+        it('Value is other error', () => {
+            const value = ErrorValueObject.create(ErrorType.NAME);
+            const valueIfNa = StringValueObject.create('error');
+            const result = textFunction.calculate(value, valueIfNa);
+            expect(result.getValue()).toBe(ErrorType.NAME);
+        });
+
+        it('Value is array', () => {
```
Comment (Dushusir) [inline packages/engine-formula/src/functions/logical/ifna/__tests__/index.spec.ts:51]: Need to support array formulas, like
image

**Review by Dushusir — CHANGES_REQUESTED** [review]

File: packages/engine-formula/src/functions/logical/not/index.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+
+import { ErrorType } from '../../../basics/error-type';
+import type { ArrayValueObject } from '../../../engine/value-object/array-value-object';
+import { type BaseValueObject, ErrorValueObject } from '../../../engine/value-object/base-value-object';
+import { BooleanValueObject } from '../../../engine/value-object/primitive-object';
+import { BaseFunction } from '../../base-function';
+
+export class Not extends BaseFunction {
```
Comment (Dushusir) [inline packages/engine-formula/src/functions/logical/not/index.ts:23]: Need to support array formulas
image

File: packages/engine-formula/src/functions/logical/switch/index.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import type { ArrayValueObject } from '../../../engine/value-object/array-value-object';
+import { ErrorValueObject } from '../../../engine/value-object/base-value-object';
+import { BaseFunction } from '../../base-function';
+
+export class Switch extends BaseFunction {
+    override minParams = 3;
+
+    override maxParams = 255;
+
+    override calculate(expression: BaseValueObject, ...casesAndResults: BaseValueObject[]): BaseValueObject {
```
Comment (Dushusir) [inline packages/engine-formula/src/functions/logical/switch/index.ts:28]: Need to support array formulas, like
image

File: packages/engine-formula/src/functions/logical/xor/index.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+import { ErrorType } from '../../../basics/error-type';
+import type { ArrayValueObject } from '../../../engine/value-object/array-value-object';
+import { type BaseValueObject, ErrorValueObject } from '../../../engine/value-object/base-value-object';
+import { BooleanValueObject, type StringValueObject } from '../../../engine/value-object/primitive-object';
+import { BaseFunction } from '../../base-function';
+
+export class Xor extends BaseFunction {
+    override minParams = 1;
```
Comment (Dushusir) [inline packages/engine-formula/src/functions/logical/xor/index.ts:24]: `=XOR(C1:C3>0,E1:G1<10,J1:J4)` gets `TRUE`
image

**Comment by Dushusir** [comment]
Please pay attention to the failure CI

File: packages/engine-formula/src/functions/logical/xor/index.ts (hunk 4 above)
Comment (Dushusir) [reply packages/engine-formula/src/functions/logical/xor/index.ts:24]: @Hossein-Khan  Still need to be resolved

**Comment by zhaolixin7** [comment]
ifna:1、not found不带引号也可以算出结果
image

**Comment by zhaolixin7** [comment]
ifna：填充后计算错
[2024_7_6 15_32_52 video.webm](https://github.com/dream-num/univer/assets/154207126/ca50c7c8-649b-4e99-aa39-e25f1a53ff86)
image

**Comment by zhaolixin7** [comment]
ifna：引用空选区，结果计算错误
image

**Comment by zhaolixin7** [comment]
ifna1、公式计算结果错误（）
image

File: packages/engine-formula/src/functions/logical/xor/index.ts (hunk 4 above)
Comment (Hossein-Khan) [reply packages/engine-formula/src/functions/logical/xor/index.ts:24]: Sure, could you please explain more about what the issue is with the XOR function?

**Comment by zhaolixin7** [comment]
not 1、引用选区单元格个为空时计算不出结果，参照Excel
image
image

**Comment by zhaolixin7** [comment]
switch 1、参数是选区时结果预期是选区，像Excel的
image
2、公式计算结果错误
image

**Comment by Dushusir** [comment]
issues move to https://github.com/dream-num/univer-pro/issues/1651 @zhaolixin7

**Pull request merged**